cognite/client/_http_client.py

```diff
@@ -110,23 +110,28 @@ def __init__(
     def request(self, method: str, url: str, **kwargs: Any) -> requests.Response:
         retry_tracker = self.retry_tracker_factory(self.config)
         headers = kwargs.get("headers")
-        last_status = None
         while True:
             try:
                 res = self._do_request(method=method, url=url, **kwargs)
-                last_status = res.status_code
+                # Cache .json() return value in order to avoid redecoding JSON if called multiple times
+                res.json = functools.lru_cache(maxsize=1)(res.json)  # type: ignore[assignment]
                 retry_tracker.status += 1
-                if not retry_tracker.should_retry(status_code=last_status):
-                    # Cache .json() return value in order to avoid redecoding JSON if called multiple times
-                    res.json = functools.lru_cache(maxsize=1)(res.json)  # type: ignore[assignment]
+
+                try:
+                    is_auto_retryable = res.json().get("error", {}).get("isAutoRetryable", False)
+                except Exception:
+                    # if the response is not JSON or it doesn't conform to the api design guide,
+                    # we assume it's not auto-retryable
+                    is_auto_retryable = False
+                if not retry_tracker.should_retry(status_code=res.status_code, is_auto_retryable=is_auto_retryable):
                     return res
             except CogniteReadTimeout as e:
                 retry_tracker.read += 1
-                if not retry_tracker.should_retry(status_code=last_status):
+                if not retry_tracker.should_retry(status_code=None, is_auto_retryable=True):
```

Comment:
Isn't this a change of behavior? It used to retry depending on the latest error, while now it will always retry?

Reply:
Shouldn't be any change in behaviour, but the code used to be a little misleading. 

It would use the latest status code yes, but that status code will always be retryable, otherwise an error would have already been raised. So in the context of a read timeout/connection error, status_code may aswell always be set to null here.